[improvement] local copy and export for images

steveruizok opened the pull request:
This PR adds a local solution for image copy and export.

In the original implementation of image export, we would send data from the local file to a serverless endpoint (https://tldraw.com/api/export), where the server would load the data into a new tldraw project, take a screenshot, and send back the resulting image. This had a few downsides: it did not work offline, took some time if the endpoint was cold, and sometimes failed altogether.

In this PR, we create the images locally. First, we generate an SVG for the selected items. Next, we load an image that contains that SVG and copy its data onto an HTML canvas. Finally, we get the image data from that canvas as a blob, which we can then copy to the clipboard or download as a file.

## Breaking Changes

### tldraw-assets.json

In order for local images (pngs, jpgs, etc) to work, we need to create an SVG with serialized versions of the app's web fonts. These are not needed when copying regular SVGs. Rather than include this data in the package itself, I've opted to include a separate `tldraw-assets.json` file, which can be found at the root of the `tldraw` repository, at `packages/tldraw-assets.json`. 

This file must be in scope in order for image assets to work correctly.

**Should the contents of this file ever change, it must be changed in each package that includes the file**, such as the VSCode editor.

### API changes

Several parts of the `TldrawApp` API have been removed, including:
- `exportAllShapesAs`
- `exportSelectedShapesAs`
- `exportSelectedShape`

The `onExport` callback is modified to contain only the image name, type, and blob.

```ts
// example/src/export.tsx

export default function Export(): JSX.Element {
  const handleExport = React.useCallback(async (info: TDExport) => {
    // When a user exports, the default behavior is to download
    // the exported data as a file. If the onExport callback is
    // provided, it will be called instead.

    const blobUrl = URL.createObjectURL(info.blob)
    const link = document.createElement('a')
    link.href = blobUrl
    // Here we open the image in a new tab rather than downloading it.
    link.target = '_blank'
    // link.download = info.name + '.' + info.type
    link.click()
  }, [])

  return (
    <div className="tldraw">
      <Tldraw onExport={handleExport} />
    </div>
  )
}
```

The `paste` method now accepts a Clipboard event as its first parameter. The app will look for files on the clipboard event's data.

Proful commented:
> First, we generate an SVG for the selected items.

Is it going to export Rich text?

What are the limitation in this approach?

Proful commented on packages/tldraw/package.json:
```diff
@@ -74,7 +76,8 @@
     "mobx": "^6.3.8",
     "react": ">=16.8",
-    "react-dom": "^16.8 || ^17.0"
+    "react-dom": "^16.8 || ^17.0",
```
Any planning to upgrade to 18 :)

steveruizok replied:
We'll have to look at this too! I haven't looked into it, so I can't say for sure whether tldraw will work correctly on React 18.

steveruizok replied:
I sure hope so though.

Proful commented on packages/tldraw/src/hooks/useStylesheet.ts:
```diff
@@ -2,18 +2,16 @@ import * as React from 'react'
 
-const UID = `Tldraw-fonts`
+const UID = `tldraw-fonts`
 const WEBFONT_URL =
```
If this can be Array & we should allow loading multiple font that is better.

steveruizok replied:
I'd like to have it all come in via CSS alone, but unfortunately we need to know a lot about font styles (in JavaScript) in order to make text measurement work. Custom fonts would need to be a separate issue but worth investigating.

steveruizok commented:
> > First, we generate an SVG for the selected items.
> 
> Is it going to export Rich text?
> 
> What are the limitation in this approach?

It's no different than the current approach in terms of generating SVGs. If you've forked tldraw and introduced shapes that can't be directly converted to SVG, then you'd still need to use a server approach. I've added that code back into the example project so that folks can see how it would be done.

steveruizok commented:
This PR also makes a few minor changes to `onAssetCreate` and `onAssetDestroy`.

judicaelandria commented:
waiting for this to get merged, my last pr will have a big change too after this
> my pr is about embedding tldraw file to github readme (exporting the selected page to png and then returning the exported image to be rendered but since you removed puppeteer and chrome for taking the screenshot, need a completely different approach to do it now, and since you don't want to use tldraw.com/export path anymore)
